src/Compilers/CSharp/Portable/Symbols/Synthesized/ReadOnlyListType/SynthesizedReadOnlyListTypeSymbol.cs

```diff
@@ -429,43 +466,45 @@ private SynthesizedReadOnlyListTypeSymbol(SourceModuleSymbol containingModule, s
                     generateNotSupportedException));
             _members = membersBuilder.ToImmutableAndFree();
 
-            // IEnumerable.GetEnumerator()
+            // IEnumerable.GetEnumerator(), IEnumerable<T>.GetEnumerator()
             static BoundStatement generateGetEnumerator(SyntheticBoundNodeFactory f, MethodSymbol method, MethodSymbol interfaceMethod)
             {
                 var containingType = (SynthesizedReadOnlyListTypeSymbol)method.ContainingType;
                 var field = containingType._field;
                 var fieldReference = f.Field(f.This(), field);
-                // return _items.GetEnumerator();
-                return f.Return(
-                    f.Call(
-                        f.Convert(
-                            interfaceMethod.ContainingType,
-                            fieldReference),
-                        interfaceMethod));
-            }
-
-            // IEnumerable<T>.GetEnumerator()
-            static BoundStatement generateGetEnumeratorT(SyntheticBoundNodeFactory f, MethodSymbol method, MethodSymbol interfaceMethod)
-            {
-                var containingType = (SynthesizedReadOnlyListTypeSymbol)method.ContainingType;
-                var field = containingType._field;
-                var fieldReference = f.Field(f.This(), field);
-                // return _items.GetEnumerator();
-                return f.Return(
-                    f.Call(
-                        f.Convert(
-                            interfaceMethod.ContainingType,
-                            fieldReference),
-                        interfaceMethod));
+                if (containingType.IsSingleton)
+                {
+                    var enumeratorType = containingType._enumeratorType;
+                    Debug.Assert(enumeratorType is not null);
+                    // return new Enumerator(_item);
+                    return f.Return(f.New(enumeratorType, fieldReference));
```

Comment:
>f.New(enumeratorType, fieldReference) [](http://example.com/codeflow?start=36&length=37)

Do we need to convert the result to `interfaceMethod.ReturnType`?

Reply:
This is a reference conversion. My understanding is that those don't need to be present in the lowered-form bound tree.